### ISSUE TITLE
Normalization of Regularization?

### ISSUE THREAD

**Issue by Laurae2** [Microsoft/LightGBM#1245]
In most frameworks we are using regularization which is automatically scaled to the number of features we have. There are rare cases where the training using non-normalized weights are causing issues, such as:

* Not having direct access to the server which trains a model on significantly more data than locally
* No direct way to query the number of observations to train
* Unknown/unspecified number of observations for the future training data

It causes a deterioration of performance of LightGBM, especially when regularization is a required element for getting good performance.

Example, using AUC, where Reg=Regularization Weight=scaled regularization:

| Number of observations | Weighted Reg | Unweighted Reg | No Regularization |
| ---: | ---: | ---: | ---: |
| 1,000 | 0.765 | 0.765 | 0.714 |
| 10,000 | 0.812 | 0.797 | 0.728 |
| 100,000 | 0.843 | 0.788 | 0.739 |
| 1,000,000 | 0.854 | 0.759 | 0.747 |
| 10,000,000 | 0.861 | 0.745 | 0.749 |

In this case, a multiplication of 1,000 in the number of observations dropped the performance of the regularization close to as if there were no regularization (and by 10,000, identical).

xgboost recently implemented weighted regularization using the number of observations to have a consistent regularization: https://github.com/dmlc/xgboost/pull/3103

**Comment by guolinke** [comment]
@Laurae2 it seems the xgboost's pr is for linear model, and LightGBM doesn't have linear models.
can you explain the idea with more details ? And where did you get the comparison table ?

**Comment by Laurae2** [comment]
> And where did you get the comparison table ?

I'm using a private dataset which has very noisy features.

> can you explain the idea with more details ?

Instead of doing regularization with absolute scaled values, regularization is set depending on the number of training observations to scale the gain/leaf value.

For instance, the gain and leaf calculation in GBDT should be the following by default:

* value = [ ((sum(gradients) - L1)² / (sum(hessians) + L2)] (with sign for leaf value)

I wonder if it makes sense to have the following instead:

* value = [ ((sum(gradients - L1))² / (sum(hessians + L2))] (with sign for leaf value)

As the sum of gradients/hessians are supposed to grow bigger with the number of observations, the impact of L1/L2 regularization is smaller. Issue is that moving L2 inside the sum means we are not directly assessing anymore the tree structure, it also breaks the maths behind boosting.

For linear regression this is fine but for GBDT it seems not directly possible. I tried some modifications to force the L1/L2 inside the sums and results were worse in general.

**Comment by guolinke** [comment]
thanks @Laurae2 , got it. 
I am aware about this problem, however, I think user should know the regularization of L1/L2 will be weaken when using large samples. 

If we want it to be auto adaptive, we will have following new problems:
1. how do we scale the L1/L2 with number of samples ? 
2. is that scale really needed ? or is regularization needed when #samples is large ?

**Comment by Laurae2** [comment]
@guolinke I'll be running a large experiment next week on Higgs to assess how regularization behaves when the number of observation changes (along with hyperparameter tuning). We might get practical hypotheses on a proper scaling from it.

**Comment by Laurae2** [comment]
@guolinke L1, L2, and hessian regularization does not seem to scale linear/log with the number of observation (although the optimal value increases with the number of observations), I think it's good for only for GLM-type models and not tree-based models.

No conclusion so far.
